pkg/utils/reconciliationerrorhandler/error_handling.go

```diff
@@ -0,0 +1,136 @@
+package reconciliationerrorhandler
+
+import (
+	"fmt"
+)
+
+// ErrorIDer is an error that may contain an ID
+type ErrorIDer interface {
+	ErrorID() string
+}
+
+// GetID returns the ID of the error if possible
+func GetID(err error) string {
+	var id string
+	if errWithID, ok := err.(ErrorIDer); ok {
+		id = errWithID.ErrorID()
+	}
+	return id
+}
+
+// reconciliationError implements ErrorIDer and Causer
+type reconciliationError struct {
+	error
+	errorID string
+}
+
+// WithID annotates the error with the given errorID which can afterwards be retrieved by ErrorID()
+func WithID(errorID string, err error) error {
+	return &reconciliationError{err, errorID}
+}
+
+func (t *reconciliationError) ErrorID() string {
+	return t.errorID
+}
+
+func (t *reconciliationError) Cause() error {
+	return t.error
+}
+
+// The ErrorContext holds the lastError IDs from the previous reconciliaton and the IDs of the errors that are processed in this context during the current reconciliation
+type ErrorContext struct {
+	name         string
+	lastErrorIDs []string
+	errorIDs     map[string]struct{}
+}
+
+// NewErrorContext creates a new error context with the given name and lastErrors from the previous reconciliation
+func NewErrorContext(name string, lastErrorIDs []string) *ErrorContext {
+	return &ErrorContext{
+		name:         name,
+		lastErrorIDs: lastErrorIDs,
+		errorIDs:     map[string]struct{}{},
+	}
+}
+
+// AddErrorID adds an error ID which will be tracked by the context and panics if more than one error have the same ID
+func (e *ErrorContext) AddErrorID(errorID string) {
+	if e.HasErrorWithID(errorID) {
+		panic(fmt.Sprintf("Error with id %q already exists in error context %q", errorID, e.name))
```

Comment:
### Question
Can this `panic` cause the GCM process to exit with error?

Reply:
The idea here is similar to the task additions in the graph which is used to generate the flow: https://github.com/gardener/gardener/blob/1ed7cdbbbbeb739b91fffb4f5c7d586c61a2815d/pkg/utils/flow/graph.go#L70

The difference is that AddErrorID is called before the execution of each task and not when ToExecute (which creates the task func) is called.

Reply:
:+1: